tempesta_fw/http.c

```diff
@@ -1121,32 +1750,37 @@ tfw_http_req_process(TfwConnection *conn, struct sk_buff *skb, unsigned int off)
 static void
 tfw_http_resp_cache_cb(TfwHttpReq *req, TfwHttpResp *resp)
 {
+	TFW_DBG2("%s: req = %p, resp = %p\n", __func__, req, resp);
 	/*
 	 * Typically we're at a node far from the node where @resp was
 	 * received, so we do an inter-node transfer. However, this is
 	 * the final place where the response will be stored. Upcoming
 	 * requests will get responded to by the current node without
 	 * inter-node data transfers. (see tfw_http_req_cache_cb())
 	 */
-	if (tfw_http_adjust_resp(resp, req))
-		goto err;
-
-	if (tfw_cli_conn_send(req->conn, (TfwMsg *)resp))
-		goto err;
-
-	TFW_INC_STAT_BH(serv.msgs_forwarded);
-	if (resp->conn)
+	if (tfw_http_adjust_resp(resp, req)) {
+		tfw_http_conn_msg_free((TfwHttpMsg *)resp);
+		tfw_http_send_500(req);
+		TFW_INC_STAT_BH(serv.msgs_otherr);
+		return;
+	}
+	/*
+	 * Responses from cache don't have @resp->conn.
+	 *
+	 * FIXME: The same check is performed in tfw_http_popreq()
+	 * which happens just a bit earlier. Is there a way to avoid
+	 * it here? The condition is considered rare, and there's no
+	 * need to check for it in the regular path. The real issue
+	 * here is that APM stats can't handle response times that
+	 * are >= USHORT_MAX. So for now don't count the requests
+	 * that are re-sent after a server connection is restored.
```

Comment:
I didn't get what the issue is. APM count delays for particular server. So if we reschedule a request to other server (or ever different connection of the server), then we should use the new, current, request timestamp for APM accounting.

TBD. Probably APM should treat connection error on special way. I.e. if the server drops connections and/or doesn't respond to sent requests at all, then we should adjust some different counters. Meantime I think we can just add maximum possible delay (probably compile-time constant) for request which were sent, but never replied.